onnxruntime/core/optimizer/bias_softmax_fusion.cc

```diff
@@ -0,0 +1,226 @@
+       (If input and mask shape match, B = k and no broadcast occurs.)
+
+       In the BERT case scores shape = [batch_size, num_heads, seq_length, seq_length]
+             and sequence mask shape = [batch_size,     1,         1,      seq_length]
+    */
+
+    // confirm all dimensions starting from softmax axis match for input and mask
+    bool singlebatch_shape_matches = true;
+
+    int axis = 1;
+    auto& softmax_attr = softmax_node.GetAttributes();
+    if (softmax_attr.find("axis") != softmax_attr.end()) {
+      auto& axis_attr = softmax_attr.at("axis");
+      axis = utils::HasInt(axis_attr)? (int)axis_attr.i() : 1;
+    }
+
+    int N1 = input1->Shape()->dim_size();
+    int N2 = input2->Shape()->dim_size();
+    int k = (int)HandleNegativeAxis(axis, std::max({N1, N2}));
+    int singlebatch_rank = std::max({N1-k, N2-k});
+
+    if (singlebatch_rank > N1 || singlebatch_rank > N2) {
+      continue;
+    }
+    for (int i = 1; i <= singlebatch_rank; i++) {
+      if (input1->Shape()->dim(N1-i) != input2->Shape()->dim(N2-i)) {
+        singlebatch_shape_matches = false;
+        break;
+      }
+    }
+
+    if (!singlebatch_shape_matches) {
+      continue;
+    }
+
+    // identify broadcast dimensions (i.e. B to k-1 in expression above)
+    // also distinguish between input and mask in this process
+    bool mask_can_simple_broadcast = true;
+
+    int B;
+    NodeArg *input, *mask;
+
+     // case 1: mask rank == input rank
+    if (N1 == N2) {
+
+      // discover B (first axis where shapes don't match)
+      B = 0;
+      while (B < k && input1->Shape()->dim(B) == input2->Shape()->dim(B)) {
+        B++;
+      }
+
+      // use B dimension to distinguish input and mask
+      select_input_on_lhs_condition(input1->Shape()->dim(B) != 1, add_node, &input, &mask);
+
+      // confirm mask dimensions are ones on broadcast axes B to (k-1)
+      for (int i = B; i < k; i++) {
+        if (mask->Shape()->dim(i) != 1) {
+          mask_can_simple_broadcast = false;
+          break;
+        }
+      }
+    }
+
+    // case 2: mask rank < input rank
+    else { 
+
+      B = 0;
+      select_input_on_lhs_condition(N1 > N2, add_node, &input, &mask);
+
+      // confirm any mask dimensions are ones before softmax axis
+      int mask_rank = mask->Shape()->dim_size();
+      for (int i = 0; i < mask_rank - singlebatch_rank; i++) {
+        if (mask->Shape()->dim(i) != 1) {
+          mask_can_simple_broadcast = false;
+          break;
+        }
+      }
+    }
+
+    if (!mask_can_simple_broadcast) {
+      continue;
+    }
+
+    // coalesce subgraph nodes into fused node
+    // -----------------------------------------------------------------------
+    std::vector<NodeArg*> fused_inputs{input, mask};
+    ILOGF("Fusing subgraph into BiasSoftmax node.\n");
+
+    std::string op_type = "BiasSoftmax";
+    Node& fused_node = graph.AddNode(graph.GenerateNodeName(op_type),
```

Comment:
can we include the name(s) of at least one of the fused nodes in the resulting node's name or description?